nutils/solver.py

```diff
@@ -180,6 +181,271 @@ def __call__(self, res0, dres0, res1, dres1):
         return min(max(scale, self.minscale), self.maxscale), scale >= self.acceptscale
 
 
+# SYSTEM
+
+class BaseSystem:
+    '''Matrix, vector, value triplet.
+
+    The BaseSystem class represents a generic evaluable block matrix, block
+    vector, and scalar value triplet. It does not make any assumptions about
+    the relation between the three objects, and it does not offer a solution
+    method, but it does provide building blocks to develop one.
+
+    Most users will want to use System instead, which extends BaseSystem with a
+    practical constructor and a suite of solution methods.
+    '''
+
+    def __init__(self, trials: Tuple[str,...], block_matrix: Tuple[Tuple[evaluable.Array,...],...], block_vector: Tuple[evaluable.Array,...], value: evaluable.Array):
+        self.trials = tuple(trials)
+        self.dtype = block_vector[0].dtype
+        if not all(v.dtype == self.dtype for v in block_vector) or not all(m.dtype == self.dtype for row in block_matrix for m in row):
+            raise ValueError('inconsistent data types in block matrix, vector')
+        _shapes = _dict((arg.name, arg.shape) for vec in block_vector for arg in vec.arguments if isinstance(arg, evaluable.Argument))
+        self.argshapes = dict(zip(trials, evaluable.compile(tuple(_shapes[t] for t in trials))()))
+        self.__argsizes = {t: numpy.prod(shape, dtype=int) for t, shape in self.argshapes.items()}
+        self.__eval = evaluable.compile((tuple(tuple(map(evaluable.as_csr, row)) for row in block_matrix), tuple(block_vector), value))
+        self.is_linear = not any(arg.name in trials for row in block_matrix for col in row for arg in col.arguments)
+        self.is_constant = self.is_linear and not any(col.arguments for row in block_matrix for col in row)
+        self.__matrix_cache = None
+
+    @functools.cached_property
+    def __nutils_hash__(self):
+        from linecache import cache # we abuse the line cache to hash self.eval
+        return types.nutils_hash((self.trials, ''.join(cache[self.__eval.__code__.co_filename][2])))
```

Comment:
I think we can add a hash to the return value of `evaluable.compile`. Then we don't have to fetch the code from the line cache, because the code is generated there.